File: LeetCode/30. Substring with Concatenation of All Words/30.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (words.length == 0 || s.length() == 0) return result;

        int wordLength = words[0].length();
        int wordCount = words.length;
        int windowLength = wordLength * wordCount;
        
        if (windowLength > s.length()) return result;

        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words) {
            wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);
        }

        // Check each possible starting point
        for (int i = 0; i < wordLength; i++) {
            int left = i;
            Map<String, Integer> currentMap = new HashMap<>();
            int count = 0;
            for (int j = i; j <= s.length() - wordLength; j += wordLength) {
                String currentWord = s.substring(j, j + wordLength);
                
                // If it is a valid word
                if (wordMap.containsKey(currentWord)) {
                    currentMap.put(currentWord, currentMap.getOrDefault(currentWord, 0) + 1);
                    count++;
                    
                    // If there are too many occurrences of a word, slide the window
                    while (currentMap.get(currentWord) > wordMap.get(currentWord)) {
                        String leftWord = s.substring(left, left + wordLength);
                        currentMap.put(leftWord, currentMap.get(leftWord) - 1);
                        count--;
                        left += wordLength;
                    }
                    
                    // If the window is valid
                    if (count == wordCount) {
                        result.add(left);
                        // Slide the window to try for the next possible solution
                        String leftWord = s.substring(left, left + wordLength);
                        currentMap.put(leftWord, currentMap.get(leftWord) - 1);
                        count--;
                        left += wordLength;
                    }
                } else {
                    // If currentWord is not a valid word
                    currentMap.clear();
                    count = 0;
                    left = j + wordLength;
                }
            }
        }
        
        return result;
    }
}

File: LeetCode/30. Substring with Concatenation of All Words/30.py
class Solution(object):
    def findSubstring(self, s, words):
        from collections import defaultdict
        
        if not s or not words:
            return []
        
        word_len = len(words[0])
        word_count = len(words)
        total_len = word_len * word_count
        word_map = defaultdict(int)
        for word in words:
            word_map[word] += 1
        
        results = []
        
        for i in range(word_len):
            left = i
            current_map = defaultdict(int)
            count = 0
            # Move right in steps of word_len
            for j in range(i, len(s) - word_len + 1, word_len):
                current_word = s[j:j + word_len]
                # Check if it is a valid word
                if current_word in word_map:
                    current_map[current_word] += 1
                    count += 1
                    
                    # If too many of one word, adjust from the left
                    while current_map[current_word] > word_map[current_word]:
                        left_word = s[left:left + word_len]
                        current_map[left_word] -= 1
                        count -= 1
                        left += word_len
                    
                    # If we have all the words in the correct counts
                    if count == word_count:
                        results.append(left)
                        left_word = s[left:left + word_len]
                        current_map[left_word] -= 1
                        count -= 1
                        left += word_len
                else:
                    # Reset if the word isn't in the dictionary
                    current_map.clear()
                    count = 0
                    left = j + word_len
        
        return results
